### PR TITLE
Rework ABI info Info structs, add some optional fields

### PR DESCRIPTION
**Pull request by lmb**
This adds the Tag field to ProgramABI, which is useful to check whether two BPF programs are in fact the same bytecode.

### CODE REVIEW

**Comment by lmb** [comment]
cc @jsitnicki

**Comment by lmb** [comment]
`/proc` doesn't have the prog tag on 4.9, so I'll need to rework this a bit.

**Comment by jsitnicki** [comment]
LGTM. Will give it another look when you have the workaround for 4.9 ready.

**Comment by lmb** [comment]
This is ready for review now. I ended up tackling #33, which has been a long standing source of confusion. The downside is that this is an API breaking change, so I plan to tag a release before merging this so that users don't get broken immediately.

@nathanjsweet @tklauser both of you have done some thinking on this before, so I'd really appreciate your input.

**Comment by lmb** [comment]
This PR is now good to go, please take a look if you have some time.

**Review by tklauser — COMMENTED** [review]
Very nice change, I think this makes the API of this package a lot easier to use.

One thing that worries me a bit are the pointer members in `type.MapInfo`/`type ProgramInfo` and that people might forget to check them for `nil` before accessing. This might work fine on the (newer) kernel they're testing their code on but would then fail e.g. when running their code on an older kernel where the field won't be set.

I'm not really sure this is a valid concern and something that could happen that often. But maybe we could think about unexporting the pointer members and adding convenience getters/setters for them instead?

File: info.go
Comment (jsitnicki) [inline info.go:75]: Can this be just a `string` that is empty when attribute is not available?

**Comment by jsitnicki** [comment]
> One thing that worries me a bit are the pointer members in `type.MapInfo`/`type ProgramInfo` and that people might forget to check them for `nil` before accessing.

You could go with Null Object Pattern and return `MapInfo` with `Type` set to `MAP_PROG_TYPE_UNSPEC` instead.

File: info.go
Comment (lmb) [reply info.go:75]: Yeah, then it's not possible to distinguish unnamed vs. no names supported. But maybe that distinction is useless in the first place.

We could do the same thing for tag, and define `""` to mean "tag is not supported" but then the naive comparison `a.Tag == b.Tag` does the wrong thing (maybe?) on older kernels where Tag is not available.

Comment (jsitnicki) [reply info.go:75]: The user has to be mindful either way. Whether its to check for `nil` pointer or for empty `string`.

**Comment by lmb** [comment]
> But maybe we could think about unexporting the pointer members and adding convenience getters/setters for them instead?

These are the signatures I could come up with:
* `func Tag(default string) string`: if there is no tag, return the default. Probably not very useful, since there isn't always a natural default, and it doesn't allow determining whether the value was present or not.
* `func Tag() (string, error)`: this is overkill, since there would only ever be a single error.
* `func Tag() (string, bool)`: this is actually quite nice, and allows ignoring the bool via `tag, _ := Tag()`. The user is forced to deal with the fact that tag might not be present.

I think I'll give the last option a try. It has the added benefit that we can avoid a bunch of allocations just for the sake of putting things behind a pointer.

**Comment by lmb** [comment]
Thank you both for your feedback, it was really valuable!

- For this PR, I've turned both Tag and Name into a plain `string`. It turns out that all kernels where we support getting `Info` (aka 4.10 and later) Tag is available. And for Name distinguishing between "has no name" and "kernel doesn't support names" doesn't seem useful.
- In the future, we should use the `func Thing() (type, bool)` for information that isn't always available.

**Review by tklauser — APPROVED** [review]
Thanks, that looks really nice now.

And I agree that `func Thing() (type, bool)` for information that isn't always available is a good solution for future changes.

**Pull request merged**
